### PR TITLE
core(lib): add error reporting

### LINKED ISSUE #2364: Non-fatal events to track
This is a list of all the non-fatal error-like events we might want to track once we have Sentry setup.

* Protocol messages with no corresponding callback
* Failed audits
* Failed gatherers
* Failed image conversions in optimize-images
* Audit debug strings for byte-efficiency, loadfast4pwa

### PR DESCRIPTION
**Pull request by patrickhulce**
fixes #2364

### CODE REVIEW

**Comment by brendankenny** [comment]
an overview of your design would be helpful (and be useful as docs for CONTRIBUTING). e.g. difference between expected errors and non-fatal ones, when to capture message vs exception, when things will be automatically collected and when it should be explicitly called (like the stack trace comment)

File: lighthouse-cli/bin.ts [hunk 1]
```diff
@@ -175,10 +176,22 @@ export async function runLighthouse(
     url: string, flags: Flags, config: Object|null): Promise<{}|void> {
   let launchedChrome: LaunchedChrome|undefined;
 
+  if (typeof flags.disableErrorReporting === 'undefined') {
```
Comment (brendankenny) [inline lighthouse-cli/bin.ts]: should be `enableErrorReporting`?

File: lighthouse-cli/bin.ts [hunk 2]
```diff
@@ -175,10 +176,22 @@ export async function runLighthouse(
...
+    flags.disableErrorReporting = !isErrorReportingEnabled;
+  }
+
   try {
     launchedChrome = await getDebuggableChrome(flags);
     flags.port = launchedChrome.port;
-    const results = await lighthouse(url, flags, config);
+    const results = await lighthouse(url, flags, config, {
+      environment: isDev() ? 'development' : 'production',
+      serverName: 'redacted',
```
Comment (brendankenny) [inline lighthouse-cli/bin.ts]: ?

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 3]
```diff
@@ -122,10 +123,17 @@ class LoadFastEnough4Pwa extends Audit {
         }
 
         if (!areLatenciesAll3G) {
+          // eslint-disable-next-line max-len
```
Comment (brendankenny) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: move not needed?

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 4]
```diff
@@ -122,10 +123,17 @@ class LoadFastEnough4Pwa extends Audit {
...
+          const debugString = `First Interactive was found at ${timeToFirstInteractive.toLocaleString()}, however, the network request latencies were not sufficiently realistic, so the performance measurements cannot be trusted.`;
+
+          Sentry.captureMessage('Network request latencies were not realistic', {
```
Comment (brendankenny) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: what is the level of detail we want to log for this sort of thing? I guess magnitude of the issue is something, and then can cross reference with the site being tested?

File: lighthouse-core/gather/driver.js [hunk 5]
```diff
@@ -287,7 +288,7 @@ class Driver {
             }
 
             // If both the data and the url are empty strings, the page had no manifest.
-            return reject('No web app manifest found.');
+            return reject(ExpectedError(new Error('No web app manifest found.')));
```
Comment (brendankenny) [inline lighthouse-core/gather/driver.js]: this seems more like impetus to add a third return possibility (e.g. `null`)

File: lighthouse-core/gather/gather-runner.js [hunk 6]
```diff
@@ -132,12 +133,15 @@ class GatherRunner {
...
+   * @param {string=} gathererName
    * @return {!Promise<*>}
    */
-  static recoverOrThrow(promise) {
+  static recoverOrThrow(promise, gathererName) {
     return promise.catch(err => {
       if (err.fatal) {
         throw err;
+      } else {
+        Sentry.captureException(err, {tags: {gatherer: gathererName}, level: 'warning'});
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: if this is just for non-fatal errors, may make more sense in `collectArtifacts` (or even `runner`)

File: lighthouse-core/gather/driver.js (hunk 5 above)
Comment (brendankenny) [inline lighthouse-core/gather/driver.js]: (and this whole method will hopefully get fixed in https://github.com/GoogleChrome/lighthouse/pull/2349#discussion_r118798623)

File: lighthouse-core/gather/gatherers/start-url.js [hunk 7]
```diff
@@ -77,7 +79,7 @@ class StartUrl extends Gatherer {
 
   afterPass(options, tracingData) {
     if (!this.startUrl) {
-      return Promise.reject(new Error('No start_url found inside the manifest'));
+      return Promise.reject(ExpectedError(new Error('No start_url found inside the manifest')));
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/start-url.js]: if we go with `ExpectedError`, it should definitely be `new ExpectedError` or `createExpectedError` and should settle on whether to pass a string or an exception in

File: lighthouse-core/lib/sentry.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
+const log = require('./log');
+
+// Fix the polyfill. See https://github.com/GoogleChrome/lighthouse/issues/73
+self.setImmediate = function (callback) {
```
Comment (brendankenny) [inline lighthouse-core/lib/sentry.js]: prefer rest args, but what did happen to the fix for #73?

File: lighthouse-core/lib/sentry.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  const args = [...arguments].slice(1);
+  Promise.resolve().then(() => callback(...args));
+  return 0;
+};
+
+const noop = () => undefined;
+const sentryDelegate = module.exports = {
```
Comment (brendankenny) [inline lighthouse-core/lib/sentry.js]: this delegate is very hard to follow but you probably already know that :)

File: lighthouse-core/runner.js [hunk 10]
```diff
@@ -33,6 +34,8 @@ class Runner {
 
     const config = opts.config;
 
+    Sentry.init(opts.flags.disableErrorReporting === false, opts.environmentData || {});
```
Comment (brendankenny) [inline lighthouse-core/runner.js]: definitely should be `opts.flags.enableErrorReporting === true`

Comment (brendankenny) [inline lighthouse-core/runner.js]: is having ` || {}` going to be confusing in the data? runner may want to try to infer more to differentiate different environments (but you might already be planning that)

**Comment by brendankenny** [comment]
and, just to be clear, the structure of things seems to be what we've been discussing for a while/what I was imaginging, so it's looking good to me :)

The biggest unfortunate thing is the expected/regular/fatal exception split, but we've also been anticipating having to go down that path for a while as well. Maybe we can revisit resolving on a sentinel value instead of rejecting on an "expected" exception, but avoid going back to the -1 world by still having them automatically managed by gather-runner/runner.

File: lighthouse-cli/bin.ts (hunk 1 above)
Comment (patrickhulce) [inline lighthouse-cli/bin.ts]: that makes opt-out (which is the only usable flag) become `--no-enable-error-reporting` which feels odd, admittedly it definitely feels odd to check `disableErrorReporting` in runner but I didn't think we had any other case where we renamed the CLI flag from CLI to core

File: lighthouse-cli/bin.ts (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-cli/bin.ts]: they send the serverName by default which is potentially private information we definitely we don't need to collect, this doesn't actually fix it but was trying it out

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 3 above)
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: ya i'll revert

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 4 above)
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: yeah saving the diff to 150ms sgtm

File: lighthouse-core/lib/sentry.js (hunk 8 above)
Comment (patrickhulce) [inline lighthouse-core/lib/sentry.js]: its still just waiting for us to bump devtools :)

File: lighthouse-core/gather/gatherers/start-url.js (hunk 7 above)
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/start-url.js]: passing a string with `createExpectedError` has the problem of sentry showing you the line where the error was created and not the more useful line here, and I thought you didn't like things extending from `Error` for some reason but I can't remember why? haha

File: lighthouse-core/gather/gather-runner.js (hunk 6 above)
Comment (patrickhulce) [inline lighthouse-core/gather/gather-runner.js]: why there? the method calling out that errors are explicitly handled was the first place that came to my mind or where this logic might live if I was looking for it, so open to much better explanations than that haha :)

File: lighthouse-core/runner.js (hunk 10 above)
Comment (patrickhulce) [inline lighthouse-core/runner.js]: mentioned above too, but why "definitely"?

File: lighthouse-cli/bin.ts [hunk 11]
```diff
@@ -151,8 +152,8 @@ function saveResults(results: Results, artifacts: Object, flags: Flags) {
         return innerPromise.then((_: Results) => Printer.write(results, outputType, outputPath));
       }, Promise.resolve(results));
     } else {
-      const outputPath =
-          flags.outputPath || `${resolvedPath}.report.${typeToExtension(flags.output)}`;
+      const outputPath = flags.outputPath ||
```
Comment (patrickhulce) [inline lighthouse-cli/bin.ts]: this was actually totally breaking vscode-insiders syntax highlighting :) filed an issue about it

File: lighthouse-cli/sentry-prompt.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+async function prompt() {
+  if (!process.stdout.isTTY || process.env.CI) {
+    // Default non-interactive sessions to false
+    return false;
+  }
+
+  let timeout: NodeJS.Timer;
+
+  const prompt = inquirer.prompt([
```
Comment (paulirish) [inline lighthouse-cli/sentry-prompt.ts]: getting an exception in the prompt() method:

image

File: lighthouse-cli/sentry-prompt.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
+import {existsSync} from 'fs';
+import {join as joinPath} from 'path';
+import {inquirer, Configstore} from './shim-modules';
+const log = require('../lighthouse-core/lib/log');
+
+const MAXIMUM_WAIT_TIME = 20 * 1000;
+
+const MESSAGE = [
+  'Lighthouse would like to report back any errors that might occur while auditing. \n  ',
+  'Information such as the URL you are auditing, its subresources, your operating system, Chrome, ',
```
Comment (paulirish) [inline lighthouse-cli/sentry-prompt.ts]: this message is pretty huge. (See screenshot) can we break it up?
fwiw the message string [can be colorized](https://github.com/yeoman/insight/blob/a0430e2cb100e45c3d073ca80c7940c165b21f65/lib/index.js#L117) before going into prompt()

> **Lighthouse requests permission to anonymously report back runtime exceptions**
> This can include data such as the test URL, its subresources, your OS, Chrome version, and Lighthouse version. 
> **May _Lighthouse_ report this data to aid in improving the tool?**

File: lighthouse-cli/sentry-prompt.ts (hunk 12 above)
Comment (patrickhulce) [inline lighthouse-cli/sentry-prompt.ts]: hmm strange, I can't repro, it happens every time?

File: lighthouse-cli/sentry-prompt.ts (hunk 13 above)
Comment (patrickhulce) [inline lighthouse-cli/sentry-prompt.ts]: sure thing

**Comment by patrickhulce** [comment]
> can you add us to the sentry project so we can see the results?

@paulirish yeah the free plan only allows one member though, time to apply for OSS account or whip out the gcard?

**Comment by paulirish** [comment]
OSS acct for sho

File: lighthouse-core/runner.js [hunk 14]
```diff
@@ -47,6 +51,19 @@ class Runner {
       return Promise.reject(err);
     }
 
+    const sentryContext = {
+      url: opts.url,
+      deviceEmulation: !opts.flags.disableDeviceEmulation,
+      networkThrottling: !opts.flags.disableNetworkThrottling,
+      cpuThrottling: !opts.flags.disableCpuThrottling,
+    };
+    Sentry.mergeContext({extra: Object.assign({}, environmentData.extra, sentryContext)});
```
Comment (paulirish) [inline lighthouse-core/runner.js]: wdyt about moving this context setting stuff into lib/sentry ?

Comment (patrickhulce) [inline lighthouse-core/runner.js]: ya done

**Comment by patrickhulce** [comment]
A not insignificant portion of this PR is now changing the behavior of the StartUrl gatherer to pass the debugString instead of an error, so might be worth breaking that out into a separate PR at this point

**Comment by brendankenny** [comment]
> so might be worth breaking that out into a separate PR at this point

I'd say 👍

File: lighthouse-cli/cli-flags.ts [hunk 15]
```diff
@@ -56,6 +56,7 @@ export function getFlags() {
           ],
           'Configuration:')
       .describe({
+        'enable-error-reporting': 'Enables error reporting (on by default)',
```
Comment (brendankenny) [inline lighthouse-cli/cli-flags.ts]: maybe we can come up with a different flag name, as it isn't really the case that it's on by default (instead it prompts by default), and a first glance may alarm some

File: lighthouse-core/gather/gather-runner.js (hunk 6 above)
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: IMO it's *just* for pulling fatal errors out of the stream. Non-fatal errors are handled on down the line (and due to a continuing lack of `fatal` errors, we should start to consider getting rid of it altogether :)

`collectArtifacts` is explicitly for collating the results of gatherers, including any errors, and has a slot ready made for [the sentry call](https://github.com/GoogleChrome/lighthouse/blob/8fce1465f376eddb88197142b5a2d0c86dc8eb66/lighthouse-core/gather/gather-runner.js#L307)

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -52,6 +52,52 @@ Don't:
...
+} catch (err) {
+  Sentry.captureException(err, {
+    tags: {audit: 'audit-name'},
+    level: 'warning',
+  });
+  doFallbackThing();
+}
+```
+
+#### If you throw an error in an audit to communicate audit failure instead of using the score property, *use createExpectedError*.
```
Comment (brendankenny) [inline CONTRIBUTING.md]: > If you throw an error in an audit to communicate audit failure instead of using the score property, *use createExpectedError*

after the changes to manifest fetching/start_url, are expected errors still needed? It sounds more like we should just discourage throwing errors to indicate an expected negative result

Comment (patrickhulce) [inline CONTRIBUTING.md]: yes I've removed createExpectedError entirely now, will update

**Comment by patrickhulce** [comment]
filed and rebased against #2549

**Comment by paulirish** [comment]
Blocked on #2549

**Comment by patrickhulce** [comment]
alright folks I think we are ready for a re-review! 🎉

**Comment by patrickhulce** [comment]
bump ⛰

File: lighthouse-cli/sentry-prompt.ts (hunk 12 above)
Comment (paulirish) [inline lighthouse-cli/sentry-prompt.ts]: `yarn install-all` fixes. :)

**Comment by paulirish** [comment]
Just ran it anew. Gave it one accidental exception and a deliberate one and its reporting well. 👍

File: lighthouse-core/lib/sentry.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+ */
+'use strict';
+
+const log = require('lighthouse-logger');
+
+// eslint-disable-next-line max-len
+const SENTRY_URL = 'https://a6bb0da87ee048cc9ae2a345fc09ab2e:63a7029f46f74265981b7e005e0f69f8@sentry.io/174697';
+
+// Fix the polyfill. See https://github.com/GoogleChrome/lighthouse/issues/73
+global.setImmediate = function(callback, ...argsForCallback) {
```
Comment (patrickhulce) [inline lighthouse-core/lib/sentry.js]: this is what actually needs the fix, so I'm in favor of leaving here until we roll to newer version

File: lighthouse-cli/cli-flags.ts [hunk 18]
```diff
@@ -104,6 +106,7 @@ Example: --output-path=./lighthouse-results.html`,
 
       // default values
       .default('chrome-flags', '')
+      .default('enable-error-reporting', true)
```
Comment (patrickhulce) [inline lighthouse-cli/cli-flags.ts]: then we won't even prompt? at that point I'd just abandon sentry entirely 😆

File: lighthouse-core/runner.js [hunk 19]
```diff
@@ -36,6 +37,13 @@ class Runner {
       return Promise.reject(err);
     }
 
+    const sentryContext = Sentry.init(opts);
+    Sentry.captureBreadcrumb({
+      message: 'Run started',
+      category: 'lifecycle',
+      data: sentryContext,
+    });
```
Comment (patrickhulce) [inline lighthouse-core/runner.js]: the point of the breadcrumb was to get easy access to the context (URL, throttling settings, etc), I'm not sure what we'd want to track at the end since we didn't have anything fatal along the way :)

File: lighthouse-cli/cli-flags.ts [hunk 20]
```diff
@@ -58,6 +58,8 @@ export function getFlags(manualArgv?: string) {
           ],
           'Configuration:')
       .describe({
+        'enable-error-reporting':
+            'Enables error reporting (prompts by default, setting this flag to false will force off error reporting).',
```
Comment (patrickhulce) [inline lighthouse-cli/cli-flags.ts]: done

File: lighthouse-cli/sentry-prompt.ts [hunk 21]
```diff
@@ -0,0 +1,66 @@
+import {existsSync} from 'fs';
```
Comment (patrickhulce) [inline lighthouse-cli/sentry-prompt.ts]: done

File: lighthouse-cli/sentry-prompt.ts [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+import {join as joinPath} from 'path';
+
+import {Configstore, inquirer} from './shim-modules';
+
+const log = require('lighthouse-logger');
+
+const MAXIMUM_WAIT_TIME = 20 * 1000;
+// clang-format off
+const MESSAGE = [
+  `${log.reset}Lighthouse is requesting permission to anonymoously report back runtime exceptions.\n  `,
```
Comment (patrickhulce) [inline lighthouse-cli/sentry-prompt.ts]: heh

File: lighthouse-cli/sentry-prompt.ts [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+  `${log.reset}This can include data such as the test URL, its subresources, your OS, Chrome version, and Lighthouse version.\n  `,
+  `May ${log.green}Lighthouse${log.reset} ${log.bold}report this data to aid in improving the tool?`,
+].join('');
```
Comment (patrickhulce) [inline lighthouse-cli/sentry-prompt.ts]: done

File: lighthouse-cli/sentry-prompt.ts [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+  if (typeof isErrorReportingEnabled === 'boolean') {
+    return isErrorReportingEnabled;
+  }
+
+  isErrorReportingEnabled = await prompt();
+  configstore.set('isErrorReportingEnabled', isErrorReportingEnabled);
+  return isErrorReportingEnabled;
+}
+
+export function isDev() {
```
Comment (patrickhulce) [inline lighthouse-cli/sentry-prompt.ts]: done

File: lighthouse-cli/test/cli/run-test.js [hunk 25]
```diff
@@ -25,6 +25,7 @@ describe('CLI run', function() {
     const url = 'chrome://version';
     const filename = path.join(process.cwd(), 'run.ts.results.json');
     const flags = getFlags(`--output=json --output-path=${filename} ${url}`);
+    flags.enableErrorReporting = false;
```
Comment (patrickhulce) [inline lighthouse-cli/test/cli/run-test.js]: otherwise the mock and expected errors from tests get reported to sentry, this test skips bin.ts where we do the prompting/setting to false for CI

File: lighthouse-cli/run.ts [hunk 26]
```diff
@@ -118,13 +118,13 @@ export function saveResults(results: Results, artifacts: Object, flags: Flags) {
 }
 
 export async function runLighthouse(
-    url: string, flags: Flags, config: Object|null): Promise<{}|void> {
+    url: string, flags: Flags, config: Object|null, environment?: Object): Promise<{}|void> {
```
Comment (patrickhulce) [inline lighthouse-cli/run.ts]: environmentData it is

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 27]
```diff
@@ -111,6 +112,20 @@ class LoadFastEnough4Pwa extends Audit {
...
 
         if (!areLatenciesAll3G) {
+          const sentryContext = Sentry.getContext();
+          const hadThrottlingEnabled = sentryContext && sentryContext.extra &&
+              sentryContext.extra.networkThrottling;
+
+          if (hadThrottlingEnabled) {
+            const violatingLatency = firstRequestLatencies
+                .find(item => Number(item.latency) < latency3gMin);
+            Sentry.captureMessage('Network request latencies were not realistic', {
```
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: yeah

File: lighthouse-cli/cli-flags.ts (hunk 18 above)
Comment (brendankenny) [inline lighthouse-cli/cli-flags.ts]: > then we won't even prompt? at that point I'd just abandon sentry entirely 😆

arg, I think this is the second time I've mixed this up :)

Taking my confusion as a sign that it's generally confusing (not necessarily true :), and since it's not only encountered reading the code, but also in `--help` as `[default: true]`, would leaving it without a default and prompting on `cliFlags.enableErrorReporting === undefined` in `bin.ts` work instead? Other ideas?

Comment (patrickhulce) [inline lighthouse-cli/cli-flags.ts]: sure that'll do

Comment (patrickhulce) [inline lighthouse-cli/cli-flags.ts]: done

File: lighthouse-core/lib/sentry.js (hunk 17 above)
Comment (brendankenny) [inline lighthouse-core/lib/sentry.js]: > until we roll to newer version

Comment (patrickhulce) [inline lighthouse-core/lib/sentry.js]: heh hey it's only been ~10 months anything can happen!

**Comment by patrickhulce** [comment]
bump ⛰

File: package.json [hunk 28]
```diff
@@ -75,13 +75,16 @@
   "dependencies": {
     "axe-core": "2.2.1",
     "chrome-devtools-frontend": "1.0.422034",
+    "configstore": "^3.1.0",
```
Comment (brendankenny) [inline package.json]: just a note so we remember, we should explore `optionalDependencies` for these (and the other shimmed ones) in the future to at least signal that these aren't necessary for folks looking for a slimmer bundle of dependencies.

**Comment by brendankenny** [comment]
PTAL

File: lighthouse-core/runner.js [hunk 29]
```diff
@@ -36,6 +37,13 @@ class Runner {
       return Promise.reject(err);
     }
 
+    const sentryContext = Sentry.init(opts);
```
Comment (patrickhulce) [inline lighthouse-core/runner.js]: Not clear what you're proposing to me since we need the options before we can start sending the errors. If you're talking about buffering the events and sending them after Sentry has been initialized that sounds like future work than initial introduction of error reporting to me :)

File: lighthouse-cli/cli-flags.ts [hunk 30]
```diff
@@ -58,6 +58,8 @@ export function getFlags(manualArgv?: string) {
           ],
           'Configuration:')
       .describe({
+        'enable-error-reporting':
+            'Enables error reporting (prompts once by default, setting this flag will force error reporting to that state).',
```
Comment (patrickhulce) [inline lighthouse-cli/cli-flags.ts]: done

File: lighthouse-core/runner.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/runner.js]: I was talking about the first thing, not buffering earlier events or anything (maybe we could eventually do that for CLI errors or something but "doesn't ever start" is the kind of thing that people would actually report manually, so probably isn't a pressing issue)

> this is mostly the earliest point we have all the available information to initialize

Does it need more than the `environmentData` made in `bin.ts` and `flags`? It seems like the `Sentry.init({environmentData, flags: cliFlags})` call could just be made there (or `lighthouse-core/index.js` might be better), but maybe I'm missing a step that's happening in between

**Comment by patrickhulce** [comment]
alright I've rebased and kicked the tires of

* timeout flow
* CI flow
* answer yes/no flow
* remembered yes/no flow
* explicit yes/no flow
* failure report flow
* extension

anything I'm missing?

File: lighthouse-cli/sentry-prompt.ts [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+import {Configstore, inquirer} from './shim-modules';
+
+const log = require('lighthouse-logger');
+
+const MAXIMUM_WAIT_TIME = 20 * 1000;
+// clang-format off
+const MESSAGE =
+  `${log.reset}Lighthouse is requesting permission to anonymously report back runtime exceptions.\n  ` +
```
Comment (paulirish) [inline lighthouse-cli/sentry-prompt.ts]: I'm reading this corp doc: https://goto.google.com/gpxxy on how to handle consent prompts..

Basically we need to clearly describe:

1. What this is about
2. What are the benefits to the user
3. What are the consequences to the user
4. What apps are affected
5. What is the name of this feature/thing and how can a user change their setting in the future
6. Where can the user learn more

There's some other UX suggestions which are pretty good, too. it's a nice read.


-------------

Let's create a `docs/error-reporting.md` which has some end-user-facing documentation on this. including:

  * how to change the setting (delete `~/.config/configstore/lighthouse.json` ?)  
  * the stuff you already put into the readme.
  * reported data includes test URL, its subresources, your OS, Chrome version, and Lighthouse version

And I'm thinking we tweak our consent opt-in message. (I got some inspiration from `yo`'s one)

Currently:

> Lighthouse is requesting permission to anonymously report back runtime exceptions.
> This can include data such as the test URL, its subresources, your OS, Chrome version, and Lighthouse version.
> May Lighthouse report this data to aid in improving the tool?

Proposed:

> We're constantly trying to improve Lighthouse and its reliability
> May we anonymously report runtime exceptions to improve the tool over time?
> Details: https://github.com/GoogleChrome/lighthouse/blob/master/docs/error-reporting.md

File: lighthouse-cli/sentry-prompt.ts [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+  `${log.reset}This can include data such as the test URL, its subresources, your OS, Chrome version, and Lighthouse version.\n  ` +
+  `May ${log.green}Lighthouse${log.reset} ${log.bold}report this data to aid in improving the tool?`;
+// clang-format on
+
+async function prompt() {
+  if (!process.stdout.isTTY || process.env.CI) {
+    // Default non-interactive sessions to false
```
Comment (paulirish) [inline lighthouse-cli/sentry-prompt.ts]: wdyt about defaulting CI to true?

Comment (patrickhulce) [reply lighthouse-cli/sentry-prompt.ts]: I'm split, I'm definitely against in *our* CI so an opt-out env variable seems good.

It'd be nice to gather errors from folks running Lighthouse in CI, but silently opting them in seems bad. Can always use `--enable-error-reporting` to skip this part entirely and run with error reporting so I'd vote no to default CI true.

File: lighthouse-cli/sentry-prompt.ts (hunk 31 above)
Comment (patrickhulce) [reply lighthouse-cli/sentry-prompt.ts]: done and done

File: lighthouse-cli/sentry-prompt.ts (hunk 32 above)
Comment (patrickhulce) [reply lighthouse-cli/sentry-prompt.ts]: I do like the idea of enabling this for our smoke tests on master though  👍  followup work

Comment (paulirish) [reply lighthouse-cli/sentry-prompt.ts]: ok sg

File: lighthouse-core/runner.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-core/runner.js]: talked to brendan about this. he doesn't like that we have to pass along this environment data from file to file, augmenting method signatures along the way. the environment data in question: https://github.com/GoogleChrome/lighthouse/blob/49c061c323fcf7767d7ef04f112fb60515f6f3ec/lighthouse-cli/bin.ts#L72-L78

i'm like "sure, fine". 🦊 

**proposal:**

* move the setup of `name`, `environment`, and `release` into runner.js. no reason why it shouldn't be defined then.
* kill `tags.channel` because hey we are only recording from the cli. deal with the channels problem later when we have channels.
* restore the method signatures of runLighthouse and runner.run etc

brendan read this and is like "cool". 🐼 

@patrickhulce wdyt?

File: docs/error-reporting.md [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+## What data gets reported?
+
+* The URL you tested
+* The runtime settings used (throttling enabled/disabled, emulation, etc)
+* The message, stack trace, and associated data of the error
+* Your Lighthouse version
+* Your Chrome version
+* Your operating system
+
+## How do I opt-in/opt-out?
```
Comment (brendankenny) [inline docs/error-reporting.md]: maybe just have "How do I opt-in?" to be clear it's only opt-in?

File: docs/error-reporting.md [hunk 34]
```diff
@@ -0,0 +1,30 @@
...
+The first time you run the CLI you will be prompted with a message asking you if Lighthouse can anonymously report runtime exceptions. You can give a direct response of `yes` or `no` to save your response, and you will not be prompted again. If no response is given within 20 seconds, a `no` response will be assumed, and you will not be prompted again. Non-interactive terminal sessions and invocations with the `CI` environment variable set will automatically be opted out.
```
Comment (brendankenny) [inline docs/error-reporting.md]: "You can give a direct response of `yes` or `no` to save your response, and you will not be prompted again." -> "You can give a direct response of `yes` or `no`, and you will not be prompted again.", maybe?

Comment (brendankenny) [reply docs/error-reporting.md]: "will automatically be opted out." -> "will automatically not be opted in." maybe? (slightly more awkward but clearer there is no state that needs to be opted out of :)

File: docs/error-reporting.md [hunk 35]
```diff
@@ -0,0 +1,30 @@
...
+
+The CLI also has two flags to control error reporting that will override the saved preference. Running Lighthouse with `--enable-error-reporting` will report errors regardless of the saved preference, and running Lighthouse with `--no-enable-error-reporting` will *not* report errors regardless of the saved preferences.
+
+## How do I change my opt-in/opt-out preference?
```
Comment (brendankenny) [inline docs/error-reporting.md]: `How do I change my opt-in preference?`?

File: docs/error-reporting.md [hunk 36]
```diff
@@ -0,0 +1,30 @@
+# Error Reporting Explained
+
+## What's going on?
+
+The Lighthouse team is constantly trying to improve the reliability of our tools. We've added functionality to anonymously report runtime exceptions, given your consent, using [Sentry](https://sentry.io/welcome/). We will use this information to detect new bugs and avoid regressions.
```
Comment (brendankenny) [inline docs/error-reporting.md]: maybe
`"If you give your consent, we've added functionality to anonymously report runtime exceptions using [Sentry](https://sentry.io/welcome/)."`

File: docs/error-reporting.md [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+
+Only CLI users are currently impacted. Node module consumers are opted out by default without a prompt, and the DevTools and extension integrations have no automatic error reporting functionality.
```
Comment (brendankenny) [inline docs/error-reporting.md]: maybe just include Node module consumers with Devtools and extension users for now since it won't be happening for them?

File: docs/error-reporting.md [hunk 38]
```diff
@@ -0,0 +1,30 @@
...
+
+## What will happen if I opt-in?
+Runtime exceptions will be reported to the team along with information on your environment such as the URL you tested, your OS, and Chrome version. See [what data gets reported](#what-data-gets-reported).
+
+## What will happen if I opt-out?
```
Comment (brendankenny) [inline docs/error-reporting.md]: `"What will happen if I don't opt-in?"`?

File: docs/error-reporting.md (hunk 34 above)
Comment (paulirish) [reply docs/error-reporting.md]: also "y" and "n" work. 
and just hitting enter defaults to "no". this should be mentioned as well.

File: docs/error-reporting.md (hunk 36 above)
Comment (patrickhulce) [reply docs/error-reporting.md]: heh, but the functionality is there if they give consent or not, the phrase is modifying "report" ;)

rephrased

File: docs/error-reporting.md (hunk 37 above)
Comment (patrickhulce) [reply docs/error-reporting.md]: yeah was trying to be explicit that the code isn't even there, but that was probably confusing

File: docs/error-reporting.md (hunk 38 above)
Comment (patrickhulce) [reply docs/error-reporting.md]: done

File: docs/error-reporting.md (hunk 33 above)
Comment (patrickhulce) [reply docs/error-reporting.md]: done

File: docs/error-reporting.md (hunk 34 above)
Comment (patrickhulce) [reply docs/error-reporting.md]: done

File: docs/error-reporting.md (hunk 35 above)
Comment (patrickhulce) [reply docs/error-reporting.md]: done

**Comment by paulirish** [comment]
@patrickhulce change lgtm

heads up: this discussion is unresolved https://github.com/GoogleChrome/lighthouse/pull/2420/files#r145858868

File: lighthouse-core/runner.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/runner.js]: kk, couple thoughts

the only signature that is modified is runLighthouse, runner.run is untouched
doing an fs.existsSync on .git directory directly in the node module and browser bundled code for a purely cli concern feels nastier than an extra optional argument, another consequence is that the package.json will also get pulled in as a dependency for the bundle when previously it was not 

totally cool with nuking tags.channel for now, but doesn't seem like it's hurting anything if we stick with passed in environmentData

if the signature disruption is a concern, how about we move Sentry.init to CLI land then (I believe this was brendan's original suggestion?)

**Comment by paulirish** [comment]
@brendankenny lets do thisssssssssss. @patrickhulce now understood what you were going for in that init() thread. :)

File: lighthouse-cli/run.ts [hunk 39]
```diff
@@ -138,13 +138,13 @@ export function saveResults(results: Results, artifacts: Object, flags: Flags) {
 }
 
 export async function runLighthouse(
-    url: string, flags: Flags, config: Object|null): Promise<{}|void> {
+    url: string, flags: Flags, config: Object|null, environmentData?: Object): Promise<{}|void> {
```
Comment (brendankenny) [inline lighthouse-cli/run.ts]: not needed anymore?

File: lighthouse-cli/bin.ts [hunk 40]
```diff
@@ -59,6 +65,23 @@ if (cliFlags.output === Printer.OutputMode[Printer.OutputMode.json] && !cliFlags
...
+      name: 'redacted', // prevent sentry from using hostname
+      environment: isDev() ? 'development' : 'production',
+      release: pkg.version,
+      tags: {
+        channel: 'cli',
+      },
+    },
+  });
+
+  return await runLighthouse(url, cliFlags, config);
```
Comment (brendankenny) [inline lighthouse-cli/bin.ts]: nit: since the function is async anyway, can drop the `await` here

Comment (patrickhulce) [reply lighthouse-cli/bin.ts]: done

File: lighthouse-cli/run.ts (hunk 39 above)
Comment (patrickhulce) [reply lighthouse-cli/run.ts]: done

**Comment by paulirish** [comment]
epic PR. super excited about this. \o/

File: lighthouse-core/runner.js (hunk 10 above)
Comment (zixuan75) [reply lighthouse-core/runner.js]: ok right
